src/Core/Domain/Country/ValueObject/CountryId.php

```diff
@@ -0,0 +1,76 @@
+<?php
+/**
+ * 2007-2019 PrestaShop and Contributors
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.txt.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Country\ValueObject;
+
+use PrestaShop\PrestaShop\Core\Domain\Country\Exception\CountryConstraintException;
+
+/**
+ * Provides country id value
+ */
+class CountryId
+{
+    const ALL_COUNTRIES_ID = 0;
+
+    /**
+     * @var int
+     */
+    private $id;
+
+    /**
+     * @param int $id
+     *
+     * @throws CountryConstraintException
+     */
+    public function __construct(int $id)
+    {
+        $this->assertPositiveInt($id);
+        $this->id = $id;
+    }
+
+    /**
+     * @return int
+     */
+    public function getValue(): int
+    {
+        return $this->id;
+    }
+
+    /**
+     * @param int $value
+     *
+     * @throws CountryConstraintException
+     */
+    private function assertPositiveInt(int $value): void
+    {
+        if (0 > $value) {
```

Comment:
If we can, it would be better indeed to use a dedicated property to carry this information. But with legacy code that uses this 0 value, we might not be able to do it 😐

Reply:
There could be a property to determine if all countries are selected but should there be another method to create this object? Static method maybe to return object with this property set and assertion of positive int skipped? There should be a check every time a value getter is called then so maybe by saying dedicated property you have something specific in mind?

Reply:
@RaimondasSapola if the "country ID zero means all countries" rule does not come from legacy and we can avoid it, let's avoid it 😄 .

In order to avoid it, the best is to be very accurate and verbose, even if it looks like a lot of code for few things.

By this, I mean: use different properties.
For example, if we can choose to add a "tax rule" for either "all countries" or "one specific country", then we need:
- one boolean property `$appliesToAllCountries`
- one integer array property `$countryId` that is nullable

And obviously, when `$appliesToAllCountries` is true then `$countryId` is null.
A bit more code, a bit more `if`, a bit more checks ... however the code becomes very simple then. Almost stupid. And this is good. We want [KISS](https://en.wikipedia.org/wiki/KISS_principle) code.